### ISSUE TITLE
Benchmark for distributed training

### ISSUE THREAD

**Issue by lee-junjie** [pytorch/pytorch#17159]
Hi, I am just wondering if there is **any** benchmark on distributed data parallel training:

For example, 
What's the speedup with **8 GPUs on 2 machines** compared to **1 GPUs on 1 machine**?
What's the speedup with **8 GPUs on 2 machines** compared to **4 GPUs on 1 machine**?

For single machine data parallel, I am also interested about the speedup, for example:
What's the speedup with **4 GPUs on 1 machines** compared to **1 GPUs on 1 machine**?

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera

**Comment by Lyken17** [comment]
Any news?

**Comment by yf225** [comment]
cc. @mrshenli Do you think we have data regarding this?

**Comment by mrshenli** [comment]
The perf numbers highly depend on the model and the hardware. 

The numbers below show the **per-GPU** training throughput on a Huggingface BERT model, where M=machine, and G=GPU per machine. (credits to @chenyangyu1988)

```
1M/1G: p50:  122/s  p75:  122/s  p90:  121/s  p95:   121/s  
1M/2G: p50:  115/s  p75:  115/s  p90:  114/s  p95:   114/s  
1M/4G: p50:  111/s  p75:  111/s  p90:  111/s  p95:   111/s  
1M/8G: p50:  110/s  p75:  109/s  p90:  108/s  p95:   107/s  
2M/8G: p50:   80/s  p75:   80/s  p90:   76/s  p95:    76/s  
```

On a single server, DDP scales almost linearly (drops ~10% when scaling from 1 GPU to 8 GPUs). With two machines, DDP uses NVLink on the same machine and Ethernet across machines. The Ethernet connection becomes the bottleneck for DDP allreduce, and the per-GPU throughput then drops considerably (~37%).

@pietern knows more on DDP perfs.

**Comment by pietern** [comment]
The speedup you can expect is a function of:
* The model size; the more weights you have, the more gradients need to be summed
* The connectivity between GPUs in a machine (NVLink, PCIe, or does everything go through the host)
* The connectivity between machines (1Gbit Ethernet or 4x100Gbit InfiniBand makes a big difference)
* The computational complexity of gradient computation; the longer it takes to compute gradients, the more time is available to spend summing gradients from different processes

Because of these factors, the only right answer here is "it depends"...

That's not to say we can't have a couple examples for widely used models of course.

**Comment by Lyken17** [comment]
@pietern I understand there are too many factors affect the final scalability. But still, it is important to have comparisons with other frameworks 

For example, horovod, they benchmark on standard model VGG / Inception / ResNet on 128 servers with 4 Pascal GPUs each connected by RoCE-capable 25 Gbit/s network

![](https://user-images.githubusercontent.com/16640218/38965607-bf5c46ca-4332-11e8-895a-b9c137e86013.png)

**Comment by pietern** [comment]
I've made a benchmark tool that can produce these numbers at #30660.

We can continue the discussion there.
